changelog.py

```diff
@@ -0,0 +1,168 @@
+import gitlab
+import os.path
+import re
+import semver
+import yaml
+
+from dateutil import parser
+from datetime import datetime, timedelta, timezone
+from jinja2 import Template
+from typing import Dict
+from urllib.parse import unquote
+
+
+path = os.path.join(os.path.dirname(__file__), "template.yml")
+with open(path) as f:
+    action = yaml.safe_load(f)
+DEFAULT_TEMPLATE = action["changelog"]["default"]
+
+DEFAULT_IGNORE = [
+    "changelog skip",
+    "duplicate",
+    "exclude from changelog",
+    "invalid",
+    "no changelog",
+    "question",
+    "wont fix",
+]
+
+
+class Changelog:
+    """A Changelog produces release notes for a single release."""
+
+    def __init__(self, repo):
+        self._repo = repo
+        self._template = Template(DEFAULT_TEMPLATE, trim_blocks=True)
+        self._ignore = set(self._slug(s) for s in DEFAULT_IGNORE)
+
+    def _slug(self, s: str) -> str:
+        """Return a version of the string that's easy to compare."""
+        return re.sub(r"[\s_-]", "", s.casefold())
+
+    def _previous_release(self, version: str):
+        """Get the release previous to the current one (according to SemVer)."""
+        cur_ver = semver.parse_version_info(version[1:])
+        prev_ver = semver.parse_version_info("0.0.0")
+        prev_rel = None
+        for tag in self._repo.tags.list(all=True):
+            if not tag.name.startswith("v"):
+                continue
+            try:
+                ver = semver.parse_version_info(tag.name[1:])
+            except ValueError:
+                continue
+            if ver.prerelease or ver.build:
+                continue
+            # Get the highest version that is not greater than the current one.
+            # That means if we're creating a backport v1.1, an already existing v2.0,
+            # despite being newer than v1.0, will not be selected.
+            if ver < cur_ver and ver > prev_ver:
+                prev_rel = tag
+                prev_ver = ver
+        return prev_rel
+
+    def _issues(self, start: datetime, end: datetime):
+        """Collect issues that were closed in the interval."""
+        issues = []
+        for x in self._repo.issues.list(state="closed", updated_after=start, all=True):
+            # If a previous release's last commit closed an issue, then that issue
+            # should be included in the previous release's changelog and not this one.
+            # The interval includes the endpoint for this same reason.
+            closed_at = parser.parse(x.closed_at)
+            if closed_at <= start or closed_at > end:
+                continue
+            if self._ignore.intersection(self._slug(label) for label in x.labels):
+                continue
+            else:
+                issues.append(x)
+        issues.reverse()  # Sort in chronological order.
+        return issues
+
+    def _merge_requests(self, start: datetime, end: datetime):
```

Comment:
There might be a nice way to make `_issues` and `_merge_requests` one function so that you aren't essentially duplicating code (since it's like 90% the same), but might not be worth it for something like this. 

Also, both these functions could use a return type

Reply:
I don't think it's worth making a method for both as enough of it would have to be separate as to not make sense

Reply:
Fair enough, probably not really worth it in this case.